gocourse02/main.go

```diff
@@ -0,0 +1,127 @@
+package main
+
+import (
+	"fmt"
+	"math/rand/v2"
+)
+
+/*
+Написати програму “Зоопарк”. Звіри повтікали (більше трьох штук), наглядач повинен їх зібрати.
+Кожна сутність (наглядач, звір, клітка, …) представляється окремою структурою (zookeeper, animal, cage, …).
+Треба використати щонайменше: структури, вказівник, nil, будування, конструктор. Додати тваринам можливість
+розмножуватись. Програма має демонструвати свою роботу через вивід в stdout.
+*/
+
+const escapedAnimalsCount = 15 // Кількість звірів які повтікали
+
+type ZooKeeper struct {
+	Name          string
+	AnimalsFound  int
+	AnimalsCaught int
+}
+
+type Catcher interface {
+	FindEscapedAnimal() bool                // Спроба знайти звіра
+	AttemptCatchAnimal(*Cage, *Animal) bool // Сброба зловити звіра в клітку
+}
+
+type Animal struct {
+	ID     int
+	Name   string
+	Gender string
+	Weight int
+}
+
+type Cage struct {
+	MaxSeats      int
+	MaxWeight     int
+	CurrentSeats  int
+	CurrentWeight int
+	Animal        []Animal
+}
+
+func NewAnimal(id, weight int) *Animal {
+	fakeData := NewFake()
+	return &Animal{ID: id + 1, Name: fakeData.Name, Gender: fakeData.Gender, Weight: weight}
+}
+
+func NewZooKeeper(name string) *ZooKeeper {
+	return &ZooKeeper{name, 0, 0}
+}
+
+func NewCage(animNumber int) *Cage {
+	numberSeats := RandInt(1, animNumber+1)    // Дамо звірям шанс
+	maxWeight := numberSeats * RandInt(10, 50) // Максимальная вага одного звіра в зоопарку 50
+	animals := make([]Animal, 0)
+	return &Cage{numberSeats, maxWeight, 0, 0, animals}
+}
+
+func (k *ZooKeeper) SearchingForAnimalsReturnToCage(catch Catcher, animal *Animal, cage *Cage) {
+	if catch.FindEscapedAnimal() {
+		k.AnimalsFound++
+		if catch.AttemptCatchAnimal(cage, animal) {
+			k.AnimalsCaught++
+			cage.CurrentSeats++
+		}
+	}
+}
+
+// Шукаємо звіра. Результат = Знаходимо або ні
+func (k *ZooKeeper) FindEscapedAnimal() bool {
+	return rand.N(2) == 1
+}
+
+// Намагаемося зловити в клітку знайденого звіра, якщо в клітці є місця  і не перевищує максимально дозволену вагу
+func (k *ZooKeeper) AttemptCatchAnimal(cage *Cage, animal *Animal) bool {
+	if cage.MaxWeight >= animal.Weight+cage.CurrentWeight && cage.MaxSeats >= cage.CurrentSeats+1 {
+		cage.CurrentWeight += animal.Weight
+		cage.Animal = append(cage.Animal, *animal)
+		return true
+	}
+	return false
+}
+
+// Функція розможноження
+func (male *Animal) Reproduction(female *Animal) {
+	avgWeight := (male.Weight + female.Weight) / 2
+	child := NewAnimal(escapedAnimalsCount, avgWeight)
+	fmt.Printf("Replay between %s and %s is complete!! Congratulations to the new animal %v \n", male.Name, female.Name, *child)
+}
+
+func RandInt(minValue, maxValue int) int {
```

Comment:
Неймінг в Go - це окреме мистецтво. Ось тут трохи можна глянути, як ідіоматично називати змінні, пакети, інтерфейси https://go.dev/talks/2014/names.slide#1

Пропоную так перейменувати параметри, можна й коментар додати:
```suggestion
// RandInt generates integer from <low, high> interval.
func RandInt(low, high int) int {
```